src/Spice86.Core/Emulator/Memory/ExpandedMemoryManager.cs

```diff
@@ -0,0 +1,777 @@
+        int destOffset = _machine.Memory.GetUint16(_machine.Cpu.State.DS, _machine.Cpu.State.SI + 14u);
+        int destPage = _machine.Memory.GetUint16(_machine.Cpu.State.DS, _machine.Cpu.State.SI + 16u);
+
+        this.SyncToEms();
+
+        if (sourceType == 0 && destType == 0) {
+            _machine.Cpu.State.AH = this.ConvToConv((uint)((sourcePage << 4) + sourceOffset), (uint)((destPage << 4) + destOffset), length);
+        } else if (sourceType != 0 && destType == 0) {
+            if (!handles.TryGetValue(sourceHandleIndex, out _)) {
+                // Return "couldn't find specified handle" code.
+                _machine.Cpu.State.AH = 0x83;
+                return;
+            }
+
+            _machine.Cpu.State.AH = this.EmsToConv(sourcePage, sourceOffset, (uint)((destPage << 4) + destOffset), length);
+        } else if (sourceType == 0 && destType != 0) {
+            if (!handles.TryGetValue(destHandleIndex, out _)) {
+                // Return "couldn't find specified handle" code.
+                _machine.Cpu.State.AH = 0x83;
+                return;
+            }
+
+            _machine.Cpu.State.AH = this.ConvToEms((uint)((sourcePage << 4) + sourceOffset), destPage, destOffset, length);
+        } else {
+            if (!handles.TryGetValue(sourceHandleIndex, out EmsHandle? sourceHandle) || !handles.TryGetValue(destHandleIndex, out EmsHandle? destHandle)) {
+                // Return "couldn't find specified handle" code.
+                _machine.Cpu.State.AH = 0x83;
+                return;
+            }
+
+            _machine.Cpu.State.AH = this.EmsToEms(sourceHandle, sourcePage, sourceOffset, destHandle, destPage, destOffset, length);
+        }
+
+        this.SyncFromEms();
+    }
+    /// <summary>
+    /// Copies data from mapped conventional memory to EMS pages.
+    /// </summary>
+    private void SyncToEms() {
+        for (int i = 0; i < MaximumPhysicalPages; i++) {
+            if (this.mappedPages[i] != -1) {
+                Span<byte> src = this.GetMappedPage(i);
+                Span<byte> dest = this.GetLogicalPage(this.mappedPages[i]);
+                src.CopyTo(dest);
+            }
+        }
+    }
+    /// <summary>
+    /// Copies data from EMS pages to mapped conventional memory.
+    /// </summary>
+    private void SyncFromEms() {
+        for (int i = 0; i < MaximumPhysicalPages; i++) {
+            if (this.mappedPages[i] != -1) {
+                Span<byte> src = this.GetLogicalPage(this.mappedPages[i]);
+                Span<byte> dest = this.GetMappedPage(i);
+                src.CopyTo(dest);
+            }
+        }
+    }
+
+    private Span<byte> GetMappedPage(int physicalPageIndex) => this._machine.Memory.Span.Slice((PageFrameSegment << 4) + (physicalPageIndex * PageSize), PageSize);
+    private Span<byte> GetLogicalPage(int logicalPageIndex) => this._machine.Memory.Span.Slice((int)this.xmsBaseAddress + (logicalPageIndex * PageSize), PageSize);
```

Comment:
Code would be way simpler overall if ems / xms pages are stored outside of memory